File: Algorithm/9/climbStairs.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
// private:
//     vector<int> memo;
//
//     int calcWays(int n){
//         if(n == 0 || n == 1)
//             return 1;
//         if(memo[n] == -1)
//             memo[n] = calcWays(n-1) + calcWays(n-2);
//         return memo[n];
//     }
// public:
//     int climbStairs(int n) {
//         memo = vector<int>(n+1, -1);
//         return calcWays(n);
//     }
public:
    int climbStairs(int n) {
        vector<int> memo(n+1, -1);
        memo[0] = 1;
        memo[1] = 1;
        for(int i = 2; i <= n; i++)
            memo[i] = memo[i-1] + memo[i-2];
        return memo[n];
    }
};

File: Algorithm/9/climbStairs.py
class Solution(object):
    def __init__(self):
        self.memo = []
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        self.memo = [-1] * (n+1)
        return self.calcWays(n)

    def calcWays(self,n):
        if n == 0 or n == 1:
            return 1
        if self.memo[n] == -1:
            self.memo[n] = self.calcWays(n-1) + self.calcWays(n-2)
        return self.memo[n]

# class Solution(object):
#     def climbStairs(self, n):
#         """
#         :type n: int
#         :rtype: int
#         """
#         if n == 1:
#             return 1
#         if n == 2:
#             return 2
#         a = 1
#         b = 2
#         res = 0
#         for i in range(3, n+1):
#             res = a + b
#             a = b
#             b = res
#         return res
